File: bindings/python/examples/host_control.py
#!/usr/bin/env python3
"""Drive the Rust kernel one step at a time, with no key or network.

Run after installing the Python binding:
    python bindings/python/examples/host_control.py
"""

from pathlib import Path
from tempfile import TemporaryDirectory

import kerness

GAMEPLAN = """---
name: host-control
agents:
  orchestrator: false
  participants: {min: 1, max: 1}
loop:
  max_turns: 4
  max_rounds: 2
  terminate_on: [DONE]
result:
  summary: {type: str}
---

Answer the host's request.
"""


class Scripted(kerness.Provider):
    def __init__(self):
        super().__init__(retries=0, backoff_sec=0)
        self.calls = 0

    def chat(self, model, messages):
        self.calls += 1
        if self.calls != 1:
            raise RuntimeError("Unexpected extra provider call")
        return kerness.ProviderResponse(
            content="Write-through keeps the invalidation rules simple.", model=model,
        )


def main():
    with TemporaryDirectory(prefix="kerness-host-control-") as directory:
        workspace = Path(directory)
        gameplan = workspace / "gameplan.md"
        gameplan.write_text(GAMEPLAN, encoding="utf-8")
        provider = Scripted()
        session = kerness.Session(
            gameplan=str(gameplan), topic="Which cache policy should we use?",
            provider=provider, model="offline-model", turn_delay_sec=0,
            memory=str(workspace / "memory.md"),
            access_policy=kerness.AccessPolicy(workspace=workspace),
        )
        session.add_agent("Advisor")
        run = session.start(
            mode="host_driven", budget={"max_provider_operations": 1},
            event_sink=lambda event: print(
                f"event {event['sequence']}: {event['event']['kind']}"
            ),
        )
        # start consumes session. Only the independent run handle is used now.
        asked = False
        while True:
            step = run.step()
            if step["status"] == "progress":
                continue
            if step["status"] != "waiting" or step["reason"]["kind"] != "input":
                raise RuntimeError(f"Unexpected step: {step}")
            if not asked:
                asked = True
                run.step({
                    "kind": "select_agent", "agent": "Advisor",
                    "instruction": "Recommend one cache policy.",
                })
                continue
            # Rust validates this shape. Finish makes no provider/judge call.
            finished = run.step({
                "kind": "finish", "result": {"summary": "Use write-through caching."},
            })
            assert finished["status"] == "finished"
            outcome = finished["outcome"]
            assert outcome["reason"]["kind"] == "completed"
            assert outcome["diagnostics"]["valid"]
            assert provider.calls == 1
            print("Result:", outcome["result"]["fields"]["summary"])
            print("Usage:", outcome["usage"])
            break


if __name__ == "__main__":
    main()

File: crates/kerness/examples/host_control.rs
//! Choose one participant, inspect events, and finish with a validated result.
//! No orchestrator, API key, network, or framework background thread is needed.
//!
//! ```sh
//! cargo run -p kerness --example host_control
//! ```

mod support;

use std::sync::Arc;
use std::time::Duration;

use kerness::access::AccessPolicy;
use kerness::usage::RunBudget;
use kerness::{
    Agent, Error, RunEvent, RunInput, RunMode, RunOptions, RunReason, Session, SessionConfig,
    StepOutcome, WaitReason,
};
use serde_json::json;

use support::{Scripted, Workspace};

const GAMEPLAN: &str = "---\nname: host-control\nagents:\n  orchestrator: false\n  \
    participants: {min: 1, max: 1}\nloop:\n  max_turns: 4\n  max_rounds: 2\n  \
    terminate_on: [DONE]\nresult:\n  summary: {type: str}\n---\n\nAnswer the host's request.\n";

fn main() -> std::result::Result<(), Box<dyn std::error::Error>> {
    let workspace = Workspace::new("host-control")?;
    let gameplan = workspace.write("gameplan.md", GAMEPLAN)?;
    let provider = Scripted::new(&["Write-through keeps the invalidation rules simple."]);
    let mut session = Session::new(SessionConfig {
        gameplan: gameplan.display().to_string(),
        topic: "Which cache policy should we use?".into(),
        provider: Some(provider.clone()),
        model: Some("offline-model".into()),
        memory: workspace.path().join("memory.md").display().to_string(),
        access_policy: Some(AccessPolicy {
            workspace: Some(workspace.path().display().to_string()),
            ..AccessPolicy::new()
        }),
        turn_delay: Duration::ZERO,
        ..SessionConfig::default()
    })?;
    session.add_agent(Agent::new("Advisor"))?;

    // start consumes configuration. The host owns this handle until completion.
    let mut run = session.start(RunOptions {
        mode: RunMode::HostDriven,
        budget: RunBudget {
            max_provider_operations: Some(1),
            ..RunBudget::default()
        },
        event_sink: Some(Arc::new(|event: &RunEvent| {
            println!("event {}: {:?}", event.sequence, event.event);
            Ok(())
        })),
        ..RunOptions::default()
    })?;
    let control = run.control(); // A UI may retain this and call cancel().
    let mut asked = false;
    loop {
        match run.step(RunInput::Continue)? {
            StepOutcome::Progress => {}
            StepOutcome::Waiting {
                reason: WaitReason::Input,
            } if !asked => {
                asked = true;
                run.step(RunInput::SelectAgent {
                    agent: "Advisor".into(),
                    instruction: "Recommend one cache policy.".into(),
                })?;
            }
            StepOutcome::Waiting {
                reason: WaitReason::Input,
            } => {
                // The host supplies the declared shape. Finish asks no model
                // for another answer or a judging pass.
                let outcome = run.step(RunInput::Finish {
                    result: json!({"summary": "Use write-through caching."}),
                })?;
                let StepOutcome::Finished { outcome } = outcome else {
                    return Err(Error::session("Finish did not complete the run").into());
                };
                assert_eq!(outcome.reason, RunReason::Completed);
                assert!(outcome.diagnostics.valid);
                assert_eq!(provider.calls(), 1);
                assert!(!control.is_cancelled());
                println!("Result: {}", outcome.result.fields["summary"]);
                println!("Usage: {}", serde_json::to_string(&outcome.usage)?);
                break;
            }
            other => return Err(Error::session(format!("Unexpected step: {other:?}")).into()),
        }
    }
    Ok(())
}
